csharp/src/SeedLang/Ast/Environment.cs

```diff
@@ -0,0 +1,79 @@
+// Copyright 2021 The Aha001 Team.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using SeedLang.Common;
+using SeedLang.Runtime;
+
+namespace SeedLang.Ast {
+  // The environment class to store scoped variables. A new frame is inserted at the first of a
+  // linked list when a new scope is entered.
+  internal class Environment {
```

Comment:
The name "environment" sounds a bit generic to me. In many cases, "environment" means the whole global environment of the runtime. Doesn't a common name like "CallStack" work here? I think the name "Call Stack" is more common when talking about the structure that holds a couple of function call frames. For example, C# has a diagnostic class named "StackFrame":

https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.stackframe?view=net-6.0

In its description: Provides information about a StackFrame, which represents a function call on the **call stack** for the current thread.

I am not that sure about this naming issue since you are using this class more like a stacked variable pool (GetVariable/SetVariable)... Just an open discussion...

Reply:
The "CallStack" will be good for a class to handle the call stack. But this class will exceed this responsibility. It only holds call frames now, but we need definitly support block-level variable scopes in the future. That is a new scope will be entered when a block is start to execute, and it will be exited when the block is ended. Then it's not just "CallStack".
Do you have any other idea about the naming of this class?

Reply:
No better name for now. Let keep the name "Environment".